### PR TITLE
Autoclose multiline strings (JEP 368)

### LINKED ISSUE #1428: Autoclose multiline strings (JEP 368)
Typing `"""` should automatically insert a new line, indent the cursor position to below the opening quotes, and finally autoclose with a triple quote
eg:
```java
String foo = """|
```
autocloses/completes to (displaying spaces):
```java
String foo = """
.............|
............."""
```

See https://openjdk.java.net/jeps/368 for more details.

@aeschli any pointers on the best way to achieve that?

### PR DESCRIPTION
**Pull request by snjeza**
Fixes #1428 

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

### CODE REVIEW

**Comment by fbricon** [comment]
@snjeza not sure about the implementation (I'll let others comment on that), but I definitely like the end-result ;-)
Now, can you address this comment (https://github.com/redhat-developer/vscode-java/issues/1428#issuecomment-622611916) about the indentation level? I just checked IntelliJ and this is what you get:
```java
        var testStringVariableForSpringJUnitMockito_WithMultiLineValues = """
                |"""
```
i.e. 2 levels of indentation after the var position.

**Comment by snjeza** [comment]
> Now, can you address this https://github.com/redhat-developer/vscode-java/issues/1428#issuecomment-622611916 about the indentation level? I just checked IntelliJ and this is what you get

I can try.
I have made it based on your request - https://github.com/redhat-developer/vscode-java/issues/1428#issue-610837274 and https://github.com/redhat-developer/vscode-java/issues/1428#issuecomment-622611916 
You can also try to format the code with the following Eclipse settings
```
org.eclipse.jdt.core.formatter.text_block_indentation=1
```
and VS Code settings
```
"editor.formatOnType": false,
```

File: language-configuration.json [hunk 1]
```diff
@@ -73,6 +84,10 @@
 		[
 			"<",
 			">"
+		],
+		[
+			"\"\"\"",
+			"\"\"\""
```
Comment (rgrunber) [inline language-configuration.json:90]: Do we need to define it for `surroundingPairs` ? It looks like for a selection, ever double-quote that's typed is handled by the `[ "\"", "\"" ]` rule so the text block never does anything. Let me know if there's a special use for this.

Comment (snjeza) [reply language-configuration.json]: You are right. I have removed it.

File: src/settings.ts [hunk 2]
```diff
@@ -287,3 +287,84 @@ function unregisterGradleWrapperPromptDialog(sha256: string) {
...
+		if (activeText === '""""""') {
+			const tabSize = <number>activeTextEditor.options.tabSize!;
+			const tabSpaces = <boolean>activeTextEditor.options.insertSpaces!;
+			let indentSize: number;
+			let repeatChar: string;
+			range = new Range(new Position(selection.line, 0), new Position(selection.line, selection.character - 2));
+			let lineText = activeTextEditor.document.getText(range);
+			lineText = replaceAll(lineText, tabSize);
+			const length = lineText.length;
+			const line = activeDocument.lineAt(selection.line);
```
Comment (rgrunber) [inline src/settings.ts:316]: Are `activeDocument` and `activeTextEditor.document` not the same thing ? I would use one consistently.

**Review by rgrunber — COMMENTED** [review]
Given the [4 options](https://git.eclipse.org/c/jdt/eclipse.jdt.core.git/tree/org.eclipse.jdt.core/formatter/org/eclipse/jdt/core/formatter/DefaultCodeFormatterConstants.java#n5817) for [`FORMATTER_TEXT_BLOCK_INDENTATION`](https://git.eclipse.org/c/jdt/eclipse.jdt.core.git/tree/org.eclipse.jdt.core/formatter/org/eclipse/jdt/core/formatter/DefaultCodeFormatterConstants.java#n2234) in JDT, I think I would prefer `INDENT_DEFAULT` (same as a wrapped line, defaulting to +2 indentations from the previous line). This is because it matches the default formatter. I think it looks a bit weird to format the text block a certain way (eg. `INDENT_ON_COLUMN`) and then to have it undone by the formatter, so we may as well keep it consistent.

I do prefer `INDENT_ON_COLUMN` though if we ever switched down the road though. @fbricon thoughts ?

File: src/extension.ts [hunk 3]
```diff
@@ -463,7 +463,8 @@ export function activate(context: ExtensionContext): Promise<ExtensionAPI> {
 						}, 100);
 					}
 				}));
-			}
+            }
+            workspace.onDidChangeTextDocument(event => onDidChangeTextDocument(event.document, event.contentChanges), null, null);
```
[the remark this answers is not in the document]
Comment (snjeza) [reply src/extension.ts]: Fixed.

File: src/settings.ts (hunk 2 above)
Comment (snjeza) [reply src/settings.ts]: Fixed.

**Comment by snjeza** [comment]
@rgrunber @fbricon I have updated the PR. See https://github.com/redhat-developer/vscode-java/pull/2320#issuecomment-1039461055

File: src/settings.ts [hunk 4]
```diff
@@ -287,3 +287,41 @@ function unregisterGradleWrapperPromptDialog(sha256: string) {
...
 	}
 }
+
+export function activateTextBlockClosing() {
+	const disposables: Disposable[] = [];
+	workspace.onDidChangeTextDocument(event => onDidChangeTextDocument(event.document, event.contentChanges), null, disposables);
+	return Disposable.from(...disposables);
+}
+
+export function onDidChangeTextDocument(document: TextDocument, changes: readonly import("vscode").TextDocumentContentChangeEvent[]): any {
```
Comment (rgrunber) [inline src/settings.ts:297]: What if you renamed this to `handleTextBlockClosing`, removed `activateTextBlockClosing` and then had :

```
context.subscriptions.push(workspace.onDidChangeTextDocument(event =>
				handleTextBlockClosing(event.document, event.contentChanges)));
```

in extension.ts .

**Review by rgrunber — COMMENTED** [review]
The change works well for me. Just a small suggestion if it can simplify the code a bit.

**Comment by snjeza** [comment]
@rgrunber I have updated the PR.

File: src/settings.ts (hunk 4 above)
Comment (snjeza) [reply src/settings.ts]: Done.

**Pull request merged**